WordPress/WordPressTest/ImageDownloaderTests.swift

```diff
@@ -68,6 +68,51 @@ class ImageDownloaderTests: CoreDataTestCase {
         }
     }
 
+    func testCancelOneOfManySubscribers() async throws {
+        // GIVEN
+        let httpRequestReceived = self.expectation(description: "HTTP request received")
+        let imageURL = try XCTUnwrap(URL(string: "https://example.files.wordpress.com/2023/09/image.jpg"))
+        stub(condition: { _ in true }, response: { _ in
+            httpRequestReceived.fulfill()
+
+            guard let sourceURL = try? XCTUnwrap(Bundle.test.url(forResource: "test-image", withExtension: "jpg")),
+                  let data = try? Data(contentsOf: sourceURL) else {
+                return HTTPStubsResponse(error: URLError(.unknown))
+            }
+
+            return HTTPStubsResponse(data: data, statusCode: 200, headers: nil)
+                .responseTime(0.3)
+        })
+
+        // WHEN there are concurrent calls to download the same image and one of those downloads is cancelled
+        let taskCompleted = self.expectation(description: "Image downloaded")
+        taskCompleted.expectedFulfillmentCount = 3
+        for _ in 1...3 {
+            try await Task.sleep(for: .milliseconds(50))
+            Task.detached {
+                do {
+                    _ = try await self.sut.image(from: imageURL)
+                } catch {
+                    XCTFail("Unexpected error: \(error)")
+                }
+                taskCompleted.fulfill()
+            }
+        }
+
+        let taskCancelled = expectation(description: "Task is cancelled")
+        let taskToBeCancelled = Task.detached {
+            do {
+                _ = try await self.sut.image(from: imageURL)
+                XCTFail("Unexpected successful result.")
+            } catch {
+                taskCancelled.fulfill()
+            }
+        }
+        taskToBeCancelled.cancel()
+
+        await fulfillment(of: [httpRequestReceived, taskCompleted, taskCancelled], timeout: 0.5)
+    }
```

Comment:
Task cancellation is cooperative. When you call [cancel()](https://developer.apple.com/documentation/swift/task/cancel()) nothing actually happens unless you do something. In this case, there are still three more tasks subscribed to the same unit of work, so the task continues as normal and finishes. It works as designed.

What does happen is `Task.isCancelled` gets set to `true`. While you can never expect `cancel()` to actually cancel the underlying work in Swift Concurrency, you can reliably use `Task.isCancelled` from the same thread to stop the work that you control from happening. For example, you can use it to stop displaying the image in the case of an image view.

```swift
let taskToBeCancelled = Task.detached {
    do {
        _ = try await self.sut.image(from: imageURL)
        XCTAssertTrue(Task.isCancelled) // succeeds
```

> .responseTime(0.3)

Relying on timing isn't ideal for testing it. If you'd like to cover this scenario with tests, I'd suggest using expectations.

Reply:
> Task cancellation is cooperative. When you call [cancel()](https://developer.apple.com/documentation/swift/task/cancel()) nothing actually happens.

I believe the cancellation would be passed along to whatever "inner" task that's currently running? In this case, if you cancel `taskToBeCancelled`, this cancellation would be passed along to the `await sut.image(for...)` task, then the `await data(for...)`, then the `await URLSession.data`.

Here is a test code:

<details>

```swift
import Foundation
import PlaygroundSupport

PlaygroundPage.current.needsIndefiniteExecution = true

func fetch() async throws -> String {
    _ = try await URLSession.shared.data(from: URL(string: "https://www.apple.com")!)
    return "Things"
}

func download() async throws -> String {
    try await fetch()
}

func use() async throws {
    _ = try await download()
}

let t = Task.detached {
    do {
        try await use()
        print("Success")
    } catch {
        print("Error: \(error)")
    }
}
t.cancel()

```

</details>

> In this case, there are still three more tasks subscribed to the same unit of work, so the download continues as normal and finishes. It works as designed.

The test fails at `XCTFail("Unexpected successful result.")`. The expectations set in the test case are: the three Task created first should finish successfully and the cancelled Task should fail with an error. If `taskToBeCancelled` is cancelled, then `taskToBeCancelled` should not return a successful result, shouldn't it?

Reply:
>Here is a test code:

This code throws the following error (`URLError`):

```swift
Error: Error Domain=NSURLErrorDomain Code=-999 "cancelled" UserInfo={NSErrorFailingURLStringKey=https://www.apple.com/, NSErrorFailingURLKey=https://www.apple.com/, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <338C255E-64EF-4162-B5EB-A0E72990C5DE>.<1>"
), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <338C255E-64EF-4162-B5EB-A0E72990C5DE>.<1>, NSLocalizedDescription=cancelled}
```

It happens because the underlying `URLSession` task listen to cancellation and handles it in this way.

In the case of the `ImageDownloader` and the test `testCancelOneOfManySubscribers`, the download doesn't get cancelled (by design, due to coalescing). So, you do call `cancel` but nothing actually gets cancelled (again, by design).

> If `taskToBeCancelled` is cancelled, then `taskToBeCancelled` should not return a successful result, shouldn't it?

I could argue both sides. The simplest implementation is to let the task finish and not throw an error. There is probably a way to update `ImageDownloader` to throw [`CancellationError`](https://developer.apple.com/documentation/swift/cancellationerror) if the task gets cancelled regardless of whether the underlying download gets cancelled or not.

Reply:
> I could argue both sides. The simplest implementation is to let the task finish and not throw an error.

Hmm... I don't know. I'm not sure how it's okay for a cancelled task/operation to return a successful result. I can understand it if there is a race condition, where the cancellation happens right at the time when the task is completed. But this test case is not that.

Besides, the `testCancellation` test expects a cancelled task returns cancelled error.

I thought the whole point of [keeping a `subscriptions` list](https://github.com/wordpress-mobile/WordPress-iOS/blob/bc3097cb5f393df086b4cbc3c00ba78ce58f7061/WordPress/Classes/Utility/Media/ImageDownloader.swift#L166) is for deciding whether to cancel the HTTP request task or not. If it's okay to not cancel the HTTP request when all "awaiter Tasks" are cancelled, is `subscriptions` related code necessary?

Reply:
This test `testCancelOneOfManySubscribers` cancels only one of four subscribers, so the download doesn't get cancelled. If you update it to cancel all four, which I did for testing, it cancels the download and finishes with `URLError` `.cancelled`. I'd suggest adding a test like that.

> understand it if there is a race condition

This always is one because you typically `cancel` from main and the download is synchronized on a background thread. So if you want to reliably stop doing any work in the UI after calling `cancel`, you need to either check `Task.isCancelled` or do it in some other way and never rely on the error thrown from a task.

I don't think there are any simple ways of updating `ImageDownloader` to throw `CancellationError` or something like that only for tasks that get canceled that don't result in the cancellation of the underlying work. It doesn't really sound right also since `cancel()` doesn't do anything in that case – the task continues, and the image gets fetched. I think it's sound.

Reply:
> This test testCancelOneOfManySubscribers cancels only one of four subscribers, so the download doesn't get cancelled.
No, the download task should not get cancelled. The `XCTFail("Unexpected error: \(error)")` assertions ensure that.

> I don't think there are any simple ways of updating ImageDownloader to throw CancellationError or something like that only for tasks that get canceled that don't result in the cancellation of the underlying work.

https://github.com/wordpress-mobile/WordPress-iOS/pull/23817/commits/dcf9997ae5852312cb012ef8d4938ca66f60b3a8 (similar to https://github.com/wordpress-mobile/WordPress-iOS/pull/23823) should fix the issue.

I'm not saying that's the only solution though. Because I thought the `_subscriptions` list is for that purpose.

Like this code here. I read it as, do not cancel if there is still subscriber waiting for the result, but cancel if there is none.

https://github.com/wordpress-mobile/WordPress-iOS/blob/e5f90677d0137a14cb6147ae37dc35329d9408cb/WordPress/Classes/Utility/Media/ImageDownloader.swift#L136-L144